File: framework/ruby/src/soda/unittest/unordered.rb
module Soda end

module Soda::Unittest

class XEntry
  def initialize(key, value)
    @key = key
    @value = value
  end

  attr_accessor :key, :value
end

class XMap
  def initialize(feature)
    @feat = feature
    @dict = Hash.new{ |hash, key| hash[key] = Array.new }
    @size = 0
  end

  def _hash(key)
    @feat.get_hash(key)
  end

  def key?(key)
    _entry(key)
  end

  def get(key, default)
    e = _entry(key)
    e ? e.value : default
  end

  def [](key)
    get(key, nil)
  end

  def []=(key, value)
    e = _entry(key)
    if e
      e.value = value
    else
      h = _hash(key)
      @dict[h] << XEntry.new(key, value)
      @size += 1
    end
  end

  def _entry(key)
    h = _hash(key)
    unless @dict.key?(h)
      return
    end
    @dict[h].each { |entry|
      if @feat.is_equal?(key, entry.key)
        return entry
      end
    }
  end

  def remove(key)
    h = _hash(key)
    unless @dict.key?(h)
      return
    end
    e = nil
    @dict[h].each { |entry|
      if @feat.is_equal?(key, entry.key)
        e = entry
        break
      end
    }
    if e
      @dict[h].delete(e)
      @size -= 1
    end
  end
end

class UnorderedListFeature
  def initialize(elem_feat)
    @elem_feat = elem_feat
  end

  def get_hash(obj)
    res = 0
    # keep low 48 bits
    mask = 0xffffffffffff
    hash_arr = obj.map { |e| @elem_feat.get_hash(e) }
    hash_arr.sort.each { |h|
      res = res * 133 + h
      res &= mask
    }
    res
  end

  def is_equal?(a, b)
    if a.size != b.size
      return false
    end
    xmap = XMap.new(@elem_feat)
    a.each { |e|
      xmap[e] = xmap.get(e, 0) + 1
    }
    b.each { |e|
      unless xmap.key?(e)
        return false
      end
      xmap[e] -= 1
      if xmap[e] == 0
        xmap.remove(e)
      end
    }
    true
  end
end

end # module Soda::Unittest

File: framework/js/src/soda/unittest/unordered.js
class XEntry {
    constructor(key, value) {
        this.key = key;
        this.value = value;
    }
}

class XMap {
    constructor(feature) {
        this.feat = feature;
        this.dict = new Map();
        this.size = 0;
    }

    _hash(key) {
        return this.feat.hash(key);
    }

    _entry(key) {
        let h = this._hash(key);
        if (!this.dict.has(h)) {
            return null;
        }
        let e = this.dict.get(h).find(x => this.feat.isEqual(key, x.key));
        return e || null;
    }

    has(key) {
        return this._entry(key) != null;
    }

    get(key, defaultVal = null) {
        let e = this._entry(key);
        return e != null ? e.value : defaultVal;
    }

    set(key, value) {
        let e = this._entry(key);
        if (e != null) {
            e.value = value;
        } else {
            let h = this._hash(key);
            if (!this.dict.has(h)) {
                this.dict.set(h, []);
            }
            this.dict.get(h).push(new XEntry(key, value));
            ++this.size;
        }
    }

    delete(key) {
        let h = this._hash(key);
        if (!this.dict.has(h)) {
            return;
        }
        let index = this.dict.get(h).findIndex(x => this.feat.isEqual(key, x.key));
        if (index >= 0) {
            this.dict.get(h).splice(index, 1);
            --this.size;
        }
    }
}

class UnorderedListFeature {
    constructor(elemFeat) {
        this.elemFeat = elemFeat;
    }

    hash(obj) {
        let res = 0;
        // keep low 48 bits
        const mask = 0xffffffffffff;
        let hashArr = obj.map(e => this.elemFeat.hash(e));
        hashArr.sort((x, y) => x - y);
        for (let h of hashArr) {
            res = res * 133 + h;
            res &= mask;
        }
        return res;
    }

    isEqual(x, y) {
        if (x.length !== y.length) {
            return false;
        }
        let xmap = new XMap(this.elemFeat);
        for (let e of x) {
            xmap.set(e, xmap.get(e, 0) + 1);
        }
        for (let e of y) {
            if (!xmap.has(e)) {
                return false;
            }
            let c = xmap.get(e) - 1;
            if (c === 0) {
                xmap.delete(e);
            } else {
                xmap.set(e, c);
            }
        }
        return true;
    }
}

exports.UnorderedListFeature = UnorderedListFeature;
